sdks/python/apache_beam/io/vcfio_test.py

```diff
@@ -0,0 +1,448 @@
+    splits = [split for split in source.split(desired_bundle_size=500)]
+    self.assertGreater(len(splits), 1)
+    sources_info = ([
+        (split.source, split.start_position, split.stop_position) for
+        split in splits])
+    self.assertGreater(len(sources_info), 1)
+    split_records = []
+    for source_info in sources_info:
+      split_records.extend(source_test_utils.read_from_source(*source_info))
+    self.assertEqual(9882, len(split_records))
+
+  def test_invalid_file(self):
+    invalid_file_contents = [
+        # Malfromed record.
+        [
+            '#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	SampleName\n',
+            '1    1  '
+        ],
+        # Missing "GT:GQ" format, but GQ is provided.
+        [
+            '#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	SampleName\n',
+            '19	123	rs12345	T	C	50	q10	AF=0.2;NS=2	GT	1|0:48'
+        ],
+        # GT is not an integer.
+        [
+            '#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	SampleName\n',
+            '19	123	rs12345	T	C	50	q10	AF=0.2;NS=2	GT	A|0'
+        ],
+        # Malformed FILTER.
+        [
+            '##FILTER=<ID=PASS,Description="All filters passed">\n',
+            '##FILTER=<ID=LowQual,Descri\n',
+            '#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	SampleName\n',
+            '19	123	rs12345	T	C	50	q10	AF=0.2;NS=2	GT:GQ	1|0:48',
+        ],
+        # Invalid Number value for INFO.
+        [
+            '##INFO=<ID=G,Number=U,Type=String,Description="InvalidNumber">\n',
+            '#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	SampleName\n',
+            '19	123	rs12345	T	C	50	q10	AF=0.2;NS=2	GT:GQ	1|0:48\n',
+        ],
+        # POS should be an integer.
+        [
+            '##FILTER=<ID=PASS,Description="All filters passed">\n',
+            '##FILTER=<ID=LowQual,Descri\n',
+            '#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	SampleName\n',
+            '19	abc	rs12345	T	C	50	q10	AF=0.2;NS=2	GT:GQ	1|0:48\n',
+        ],
+    ]
+    for content in invalid_file_contents:
+      try:
+        self._read_records(self._create_temp_vcf_file(content))
+        self.fail('Invalid VCF file must throw an exception')
+      except ValueError:
+        pass
+    # Try with multiple files (any one of them will throw an exception).
+    tmpdir = self._new_tempdir()
+    for content in invalid_file_contents:
+      self._create_temp_vcf_file(content, tmpdir=tmpdir)
+    try:
+      self._read_records(os.path.join(tmpdir, '*.vcf'))
+      self.fail('Invalid VCF file must throw an exception.')
+    except ValueError:
+      pass
+
+  def test_no_samples(self):
+    header_line = '#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO\n'
+    record_line = '19	123	.	G	A	.	PASS	AF=0.2'
+    expected_variant = Variant(
+        reference_name='19', start=122, end=123, reference_bases='G',
+        alternate_bases=['A'], filters=['PASS'],
+        info={'AF': VariantInfo(data=[0.2], field_count='A')})
```

Comment:
<!--new_thread; commit:70ade9d4fe6a8e31ff0be4d39b731510c93b4bf9; resolved:0-->
Please add tests similar to following from textio.

test_read_reentrant...()
test_dynamic_work_rebalancing...()

Reply:
Got it :+1: